### PR TITLE
Persist NEP5 Token creation 

### PR DESCRIPTION
**Pull request by localhuman**
**What current issue(s) does this address?, or what feature is it adding?**
- This PR adds functionality to listen to the `SmartContractEvent.CREATE` and then determine if that event created an NEP5 compliant token.
- If it is determined to be an NEP5, then that event is persisted alongside other notifications in the NotificationDB.  
- This makes a full list of all NEP5 Compliant tokens queryable via the `neo/api/REST/NotificationRestApi`

**How did you solve this problem?**
- code

**How did you make sure your solution works?**
- tests

**Did you add any tests?**
- yes

**Are there any special changes in the code that we should be aware of?**
- not really

**Did you run `make lint` and `make test`?**
- yes

**Are you making a PR to a feature branch or development rather than master?**
- yes

### CODE REVIEW

File: neo/Implementations/Notifications/LevelDB/NotificationDB.py [hunk 1]
```diff
@@ -166,3 +210,56 @@ def get_by_addr(self, address):
...
+        contractlist_snapshot = self.db.prefixed_db(NotificationPrefix.PREFIX_CONTRACT).snapshot()
+        results = []
+
+        for val in contractlist_snapshot.iterator(prefix=bytes(hash.Data), include_key=False):
+            if len(val) > 4:
+                try:
+                    event = SmartContractEvent.FromByteArray(val)
+                    results.append(event)
+                except Exception as e:
+                    logger.info("could not parse event: %s " % val)
```
Comment (metachris) [inline neo/Implementations/Notifications/LevelDB/NotificationDB.py]: is this an info or should it show up red like an error? Then use `logger.error` or `logger.exception`

File: neo/Implementations/Notifications/LevelDB/NotificationDB.py [hunk 2]
```diff
@@ -166,3 +210,56 @@ def get_by_addr(self, address):
...
+        """
+        tokens_snapshot = self.db.prefixed_db(NotificationPrefix.PREFIX_TOKEN).snapshot()
+
+        try:
+            val = tokens_snapshot.get(hash.ToBytes())
+            if val:
+                event = SmartContractEvent.FromByteArray(val)
+                return event
+        except Exception as e:
+            logger.info("Smart contract event with contract hash %s not found: %s " % (hash.ToString(), e))
```
Comment (metachris) [inline neo/Implementations/Notifications/LevelDB/NotificationDB.py]: is this an info or should it show up red like an error? Then use `logger.error` or `logger.exception`

File: neo/SmartContract/SmartContractEvent.py [hunk 3]
```diff
@@ -84,7 +103,17 @@ def Deserialize(self, reader):
...
+            self.contract = ContractState()
+            self.contract.Deserialize(reader)
+
+            try:
+                from neo.Wallets.NEP5Token import NEP5Token
+                token = NEP5Token(binascii.hexlify(self.contract.Code.Script))
+                token.Deserialize(reader)
+                self.token = token
+            except Exception as e:
+                print("Couldnt deserialize token %s " % e)
```
Comment (metachris) [inline neo/SmartContract/SmartContractEvent.py]: should be `logger.error` or `logger.exception`

File: neo/Wallets/NEP5Token.py [hunk 4]
```diff
@@ -91,15 +95,19 @@ def Query(self, wallet):
...
             self.symbol = results[1].GetString()
             self.decimals = results[2].GetBigInteger()
-            return True
+            if len(self.name) > 1 and self.name != 'Stack Item' \
+                    and len(self.symbol) > 1 and self.symbol != 'Stack Item'\
+                    and self.decimals < 10:
+                return True
         except Exception as e:
-            logger.error("could not query token %s " % e)
+            logger.debug("could not query token %s " % e)
```
Comment (metachris) [inline neo/Wallets/NEP5Token.py]: Can make sense. Sure?

File: protocol.testnet.json [hunk 5]
```diff
@@ -10,7 +10,7 @@
             "http://*:20332"
         ],
         "WsPort": 20334,
-        "BootstrapFile":"https://s3.us-east-2.amazonaws.com/cityofzion/bootstrap_testnet/Chain780000.tar.gz",
+        "BootstrapFile":"https://s3.us-east-2.amazonaws.com/cityofzion/bootstrap_testnet/ChainTestnet1060000.tar.gz",
```
Comment (metachris) [inline protocol.testnet.json]: 👍  ;)

File: neo/Core/State/ContractState.py [hunk 6]
```diff
@@ -128,6 +143,25 @@ def Serialize(self, writer):
         writer.WriteVarString(self.Email)
         writer.WriteVarString(self.Description)
 
+
+    def DetermineIsNEP5(self):
+
```
Comment (ixje) [inline neo/Core/State/ContractState.py]: Please add docstrings

File: neo/Implementations/Notifications/LevelDB/NotificationDB.py [hunk 7]
```diff
@@ -198,6 +212,25 @@ def get_by_addr(self, address):
                     logger.info("could not parse event: %s " % val)
         return results
 
+    def get_by_contract(self, contract_hash):
+        hash = contract_hash
```
Comment (ixje) [inline neo/Implementations/Notifications/LevelDB/NotificationDB.py]: Please add docstrings

File: neo/VM/ScriptBuilder.py [hunk 8]
```diff
@@ -190,6 +190,19 @@ def EmitAppCallWithOperation(self, script_hash, operation):
         self.push(operation.encode('utf-8').hex())
         self.Emit(APPCALL, script_hash.Data)
 
+    def EmitAppCallWithJsonArgs(self, script_hash, args):
+        args.reverse()
```
Comment (ixje) [inline neo/VM/ScriptBuilder.py]: Please add docstrings

File: neo/Wallets/NEP5Token.py [hunk 9]
```diff
@@ -277,6 +283,15 @@ def CrowdsaleRegister(self, wallet, register_addresses):
 
         return tx, fee, results
 
+    def Serialize(self, writer):
+        writer.WriteVarString(self.name)
+        writer.WriteVarString(self.symbol)
+        writer.WriteUInt8(self.decimals)
+
+    def Deserialize(self, reader):
+        self.name = reader.ReadVarString().decode('utf-8')
```
Comment (ixje) [inline neo/Wallets/NEP5Token.py]: """
        Deserialize full object.

        Args:
            reader (neo.IO.BinaryReader):
        """

File: neo/Wallets/NEP5Token.py [hunk 10]
```diff
@@ -277,6 +283,15 @@ def CrowdsaleRegister(self, wallet, register_addresses):
 
         return tx, fee, results
 
+    def Serialize(self, writer):
+        writer.WriteVarString(self.name)
```
Comment (ixje) [inline neo/Wallets/NEP5Token.py]: """
        Serialize object.

        Args:
            writer (neo.IO.BinaryWriter):
        """

Comment (localhuman) [reply neo/Wallets/NEP5Token.py]: 😆 glad the docstring police are in force

**Pull request merged**